paddle/fluid/pir/transforms/onednn/conv_elementwise_add_mkldnn_fuse_pass.cc

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2024 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/pir/transforms/onednn/conv_elementwise_add_mkldnn_fuse_pass.h"
+
+#include "paddle/fluid/pir/dialect/operator/ir/onednn_op.h"
+#include "paddle/fluid/pir/dialect/operator/ir/pd_op.h"
+#include "paddle/fluid/pir/drr/include/drr_pattern_base.h"
+#include "paddle/fluid/pir/utils/general_functions.h"
+
+#include "paddle/pir/include/pass/pass.h"
+#include "paddle/pir/include/pass/pass_registry.h"
+
+namespace {
+
+class ConvElementwiseAddPattern : public paddle::drr::DrrPatternBase {
+ private:
+  std::string conv_name_;
+  std::string fused_conv_name_;
+
+ public:
+  ConvElementwiseAddPattern(const std::string &conv_name,
+                            const std::string &fused_conv_name)
+      : conv_name_(conv_name), fused_conv_name_(fused_conv_name) {}
+
+  std::string name() const override { return "ConvElementwiseAddPattern"; }
+
+  uint32_t benefit() const override { return 2; }
+
+  void operator()(paddle::drr::DrrPatternContext *ctx) const override {
+    paddle::drr::SourcePattern pat = ctx->SourcePattern();
+
+    const auto &conv =
+        pat.Op(conv_name_,
+               {{"strides", pat.Attr("strides")},
+                {"paddings", pat.Attr("paddings")},
+                {"padding_algorithm", pat.Attr("padding_algorithm")},
+                {"dilations", pat.Attr("dilations")},
+                {"groups", pat.Attr("groups")},
+                {"data_format", pat.Attr("data_format")}});
+
+    const auto &add = pat.Op(paddle::dialect::AddOp::name());
+    conv({&pat.Tensor("input"), &pat.Tensor("filter")},
+         {&pat.Tensor("conv2d_out")});
+
+    pat.Tensor("add_out") =
+        add(pat.Tensor("conv2d_out"), pat.Tensor("residual_param"));
+    pat.RequireNativeCall(
+        [](const paddle::drr::MatchContext &match_ctx) -> bool {
+          if (!pir::ValueIsPersistable(match_ctx.Tensor("residual_param"))) {
```

Comment:
Since only weight is constant-needed, I think for elementwise_add, it's appropriate not to check the consistance of "residual_param"?